saw-remote-api/src/SAWServer/LLVMCrucibleSetup.hs

```diff
@@ -237,4 +238,5 @@ llvmLoadModule (LLVMLoadModuleParams serverName fileName) =
               Left err -> raise (cantLoadLLVMModule (LLVM.formatError err))
               Right llvmMod ->
                 do setServerVal serverName llvmMod
+                   trackFile fileName
```

Comment:
There is a race condition here. I think that avoiding it will entail changes to SAW similar to this change to Cryptol: https://github.com/GaloisInc/cryptol/pull/576

Basically, the file could change between the invocation of `llvmLoadModule` and `trackFile`. Ideally, we get a hash back from the module loading code in addition to the parsed result.

Reply:
As an intermediary measure, could we hash the file both before and after, and fail if any of the three total hashes mismatch?

Reply:
Ah, I see—I just looked at the implementation in Cryptol and I now understand why the above double-fingerprint solution would be inadequate too: if the file is changed between the first fingerprinting and the module load, and then changed *back* to its original state before the second fingerprinting, there could again be a mismatch. I'm quite hesitant to merge a potential race condition to caching code. Seems like we should implement the same functionality in SAW's LLVM modules... I'll look into what that would entail.

Reply:
The LLVM verifier [loads modules from disk like so](https://github.com/GaloisInc/llvm-verifier/blob/af5f873253db889dc7e22c172894f853c52a4218/src/Verifier/LLVM/Codebase.hs#L207-L214):

```haskell
-- | Load a module, calling fail if parsing or loading fails.
loadModule :: FilePath -> IO L.Module
loadModule bcFile = do
  eab <- CE.mapException (LLVMLoadMiscException bcFile) $
         BC.parseBitCodeFromFile bcFile
  case eab of
    Left msg  -> CE.throw (LLVMLoadParseFailure bcFile msg)
    Right mdl -> return mdl
```

In order to prevent the race condition, we would need to replace the call to `BC.parseBitCodeFromFile` (which is [`Data.LLVM.BitCode.parseBitCodeFromFile`](https://github.com/GaloisInc/llvm-pretty-bc-parser/blob/master/src/Data/LLVM/BitCode.hs#L24-L25)) with a simultaneous load/hash. Currently, `BC.parseBitCodeFromFile` is defined (in [`llvm-pretty-bc-parser`](https://github.com/GaloisInc/llvm-pretty-bc-parser)) as (some imports elided):

```haskell
import Data.LLVM.BitCode.Bitstream (Bitstream, parseBitCodeBitstream)
import qualified Data.ByteString as S

parseBitCodeFromFile :: FilePath -> IO (Either Error Module)
parseBitCodeFromFile  = parseBitCode <=< S.readFile

parseBitCode :: S.ByteString -> IO (Either Error Module)
parseBitCode  = parseBitstream . parseBitCodeBitstream

parseBitstream :: Either String Bitstream -> IO (Either Error Module)
parseBitstream e = -- ... (elided)
```

The simplest solution would be to load the whole file as a `ByteString` into memory within SAW, then hash the ByteString, then parse it into an LLVM `Module`. This would suffice to ensure that the loaded `Module` corresponds to the file contents with the hash we compute.

**A note on a possible optimization:** This would not have worse memory usage than the present situation. However, this does bring to light a possible memory usage optimization: we may not need to load the `ByteString` strictly into memory to begin with: we could stream it through both the hash function and the parser without allocating the intermediate strict `ByteString`. This refactor is not necessary to implement this change, however—I'm noting it here merely for the purpose of pointing out its possibility.

Reply:
After discussion with @david-christiansen, we've realized that this solution does not in fact remove the race condition—and indeed, there is no way to prevent such a race condition without some kind of transactional file-locking mechanism, because a file could always be changed on disk during the hash-*checking* step, after its own hash has been computed but before that hash has been compared to its previous hash. This possibility exists in the Cryptol module reloading code also, with similar caveats. As such we're punting on avoiding this race condition, and merging with the knowledge that it exists.